### ISSUE TITLE
Security for Event framework

### ISSUE THREAD

**Issue by muralisrini** [hyperledger-archives/fabric#857]
Currently events do not have any security or access control mechanisms. We need to implement security for events which goes hand-in-hand with the security mechanisms in place for transactions.

This will likely span multiple sprints. Starting with Sprint 2.

**Comment by ijmitch** [comment]
I'm very interested in this feature.

I'm looking at the events as an important way for the actions taking place inside the ledger to interact with systems and services outside of the ledger. One of my scenarios is a simple two-party transaction. If one of those parties submits the transaction, they will naturally be able to observe the outcome... but how does the other party get informed that the transaction happened?

So in that simple scenario, I would expect that as a blockchain participant I would be able to register and listen for (at least) commit events for transactions that I'm 'part of'.

Otherwise, I don't see how a non-submitter can get knowledge of transactions they're involved in.

I think that's aligned with what you've said about the 'security mechanisms in place for transactions'.

**Comment by muralisrini** [comment]
@ijmitch : that about sums it up I think.  Just to emphasize the point, allow me bring it out here.

_"..events as an important way for the actions taking place inside the ledger to interact with systems and services outside of the ledger.."_

In other words,  keep chaincode simple and deterministic, use events to build complex apps.

In order to really do that, events infrastructure should be beefed up from in two fronts (1) security and (2) chaincode events.

**Comment by ijmitch** [comment]
yes @muralisrini , we're on the same page!

**Comment by ijmitch** [comment]
So just to expand a bit having reassurance I'm on the right track...

Just as a submitter can choose a peer to submit to, as a listener I would pick a peer to register with. This resolves the potential ambiguity of the event 'happening' in all the peers and listeners receiving multiple notifications (and is nicely symmetric). I don't read section 3.5 in protocol-spec this way though.

This does make the listener vulnerable to that peer failing. This ought to be considered once we get to peer failover scenarios - as a listener I would like to inherit the resilience of the peer I register with.

When chaincode generates an event, is notification dependent on the transaction actually making it on to the ledger? ie a commit? I was imagining a common need for suitably registered and authorised listeners to simply get the notification of the committed transaction (including all the data the submitter supplied). This would be my first preference before going after custom events.

**Comment by muralisrini** [comment]
Sounds right to me...

Typically, you would register with one peer with the understanding that you could register with many based on application needs (such as failover).

Notify-generate-events-on-commit-only would fall in line with the desired properties of blockchain. That would be a first choice of implementation I'd think (as opposed to, say, "ephemeral" events).

**Comment by ijmitch** [comment]
Is there somewhere I can read about things like Notify-generate-events-on-commit-only ?

**Comment by ijmitch** [comment]
@muralisrini ... I see you're thinking about this with @chetmurthy - https://github.com/chetmurthy/fabric-wiki/blob/master/chet-arch/event-subsystem.txt

**Comment by prjayach** [comment]
Yes, we are using events the same way -- for notifying other participants who interact with the blockchain using their own applications.
